File: app/src/main/java/ca/pkay/rcloneexplorer/util/SyncServiceNotifications.java
package ca.pkay.rcloneexplorer.util;

import static ca.pkay.rcloneexplorer.Services.SyncService.EXTRA_TASK_ID;
import static ca.pkay.rcloneexplorer.Services.SyncService.TASK_ACTION;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.Build;

import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;

import java.util.ArrayList;

import ca.pkay.rcloneexplorer.BroadcastReceivers.SyncCancelAction;
import ca.pkay.rcloneexplorer.R;
import ca.pkay.rcloneexplorer.Services.SyncService;

public class SyncServiceNotifications {

    private final String OPERATION_FAILED_GROUP = "ca.pkay.rcexplorer.OPERATION_FAILED_GROUP";
    private final String OPERATION_SUCCESS_GROUP = "ca.pkay.rcexplorer.OPERATION_SUCCESS_GROUP";
    private static final String CHANNEL_ID = "ca.pkay.rcexplorer.sync_service";
    private static final String CHANNEL_NAME = "Sync service";

    public static final int PERSISTENT_NOTIFICATION_ID_FOR_SYNC = 162;
    private static final int OPERATION_FAILED_NOTIFICATION_ID = 89;
    private static final int OPERATION_SUCCESS_NOTIFICATION_ID = 698;

    Context mContext;

    public SyncServiceNotifications(Context context) {
        this.mContext = context;
    }

    public void showFailedNotification(String title, String content, int notificationId, long taskid) {

        Intent i = new Intent(mContext, SyncService.class);
        i.setAction(TASK_ACTION);
        i.putExtra(EXTRA_TASK_ID, taskid);
        PendingIntent retryPendingIntent = PendingIntent.getService(mContext, 0, i, PendingIntent.FLAG_UPDATE_CURRENT);

        NotificationCompat.Builder builder = new NotificationCompat.Builder(mContext, CHANNEL_ID)
                .setSmallIcon(R.drawable.ic_twotone_cloud_error_24)
                .setContentTitle(title)
                .setContentText(content)
                .setStyle(new NotificationCompat.BigTextStyle().bigText(content))
                .setGroup(OPERATION_FAILED_GROUP)
                .setPriority(NotificationCompat.PRIORITY_LOW)
                .addAction(R.drawable.ic_refresh, mContext.getString(R.string.retry_failed_sync), retryPendingIntent);

        NotificationManagerCompat notificationManager = NotificationManagerCompat.from(mContext);
        notificationManager.notify(notificationId, builder.build());
        createSummaryNotificationForFailed();
    }

    public void showSuccessNotification(String content, int notificationId) {

        NotificationCompat.Builder builder = new NotificationCompat.Builder(mContext, CHANNEL_ID)
                .setSmallIcon(R.drawable.ic_twotone_cloud_done_24)
                .setContentTitle(mContext.getString(R.string.operation_success))
                .setContentText(content)
                .setStyle(new NotificationCompat.BigTextStyle().bigText(content))
                .setGroup(OPERATION_SUCCESS_GROUP)
                .setPriority(NotificationCompat.PRIORITY_LOW);

        NotificationManagerCompat notificationManager = NotificationManagerCompat.from(mContext);
        notificationManager.notify(notificationId, builder.build());
        createSummaryNotificationForSuccess();
    }

    public void createSummaryNotificationForFailed() {
        Notification summaryNotification =
                new NotificationCompat.Builder(mContext, CHANNEL_ID)
                        .setContentTitle(mContext.getString(R.string.operation_failed))
                        //set content text to support devices running API level < 24
                        .setContentText(mContext.getString(R.string.operation_failed))
                        .setSmallIcon(R.drawable.ic_twotone_cloud_error_24)
                        .setGroup(OPERATION_FAILED_GROUP)
                        .setGroupSummary(true)
                        .setAutoCancel(true)
                        .build();

        NotificationManagerCompat notificationManager = NotificationManagerCompat.from(mContext);
        notificationManager.notify(OPERATION_FAILED_NOTIFICATION_ID, summaryNotification);
    }

    public void createSummaryNotificationForSuccess() {
        Notification summaryNotification =
                new NotificationCompat.Builder(mContext, CHANNEL_ID)
                        .setContentTitle(mContext.getString(R.string.operation_success))
                        //set content text to support devices running API level < 24
                        .setContentText(mContext.getString(R.string.operation_success))
                        .setSmallIcon(R.drawable.ic_twotone_cloud_done_24)
                        .setGroup(OPERATION_SUCCESS_GROUP)
                        .setGroupSummary(true)
                        .setAutoCancel(true)
                        .build();

        NotificationManagerCompat notificationManager = NotificationManagerCompat.from(mContext);
        notificationManager.notify(OPERATION_SUCCESS_NOTIFICATION_ID, summaryNotification);
    }

    public void setNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            // Create the NotificationChannel, but only on API 26+ because
            // the NotificationChannel class is new and not in the support library
            NotificationChannel channel = new NotificationChannel(CHANNEL_ID, CHANNEL_NAME, NotificationManager.IMPORTANCE_LOW);
            channel.setDescription(mContext.getString(R.string.sync_service_notification_channel_description));
            // Register the channel with the system
            NotificationManager notificationManager = (NotificationManager) mContext.getSystemService(Context.NOTIFICATION_SERVICE);
            if (notificationManager != null) {
                notificationManager.createNotificationChannel(channel);
            }
        }
    }

    public NotificationCompat.Builder getPersistentNotification(String title) {
        Intent foregroundIntent = new Intent(mContext, SyncService.class);
        PendingIntent pendingIntent = PendingIntent.getActivity(mContext, 0, foregroundIntent, 0);

        Intent cancelIntent = new Intent(mContext, SyncCancelAction.class);
        PendingIntent cancelPendingIntent = PendingIntent.getBroadcast(mContext, 0, cancelIntent, 0);

        NotificationCompat.Builder builder = new NotificationCompat.Builder(mContext, CHANNEL_ID)
                .setSmallIcon(R.drawable.ic_twotone_rounded_cloud_sync_24)
                .setContentTitle(mContext.getString(R.string.syncing_service, title))
                .setContentIntent(pendingIntent)
                .addAction(R.drawable.ic_cancel_download, mContext.getString(R.string.cancel), cancelPendingIntent)
                .setPriority(NotificationCompat.PRIORITY_LOW);

        return builder;
    }

    public void updateNotification(String title, String content, ArrayList<String> bigTextArray, int percent) {
        StringBuilder bigText = new StringBuilder();
        for (int i = 0; i < bigTextArray.size(); i++) {
            bigText.append(bigTextArray.get(i));
            if (i < 4) {
                bigText.append("\n");
            }
        }

        Intent foregroundIntent = new Intent(mContext, SyncService.class);
        PendingIntent pendingIntent = PendingIntent.getActivity(mContext, 0, foregroundIntent, 0);

        Intent cancelIntent = new Intent(mContext, SyncCancelAction.class);
        PendingIntent cancelPendingIntent = PendingIntent.getBroadcast(mContext, 0, cancelIntent, 0);

        NotificationCompat.Builder builder = new NotificationCompat.Builder(mContext, CHANNEL_ID)
                .setSmallIcon(R.drawable.ic_twotone_rounded_cloud_sync_24)
                .setContentTitle(mContext.getString(R.string.syncing_service, title))
                .setContentText(content)
                .setContentIntent(pendingIntent)
                .setStyle(new NotificationCompat.BigTextStyle().bigText(bigText.toString()))
                .addAction(R.drawable.ic_cancel_download, mContext.getString(R.string.cancel), cancelPendingIntent)
                .setPriority(NotificationCompat.PRIORITY_LOW)
                .setProgress(100, percent, false);

        NotificationManagerCompat notificationManagerCompat = NotificationManagerCompat.from(mContext);
        notificationManagerCompat.notify(PERSISTENT_NOTIFICATION_ID_FOR_SYNC, builder.build());
    }

}

File: app/src/main/java/ca/pkay/rcloneexplorer/notifications/SyncServiceNotifications.kt
package ca.pkay.rcloneexplorer.notifications

import android.app.Notification
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.preference.PreferenceManager
import androidx.work.WorkManager
import ca.pkay.rcloneexplorer.BroadcastReceivers.SyncRestartAction
import ca.pkay.rcloneexplorer.R
import ca.pkay.rcloneexplorer.util.FLog
import ca.pkay.rcloneexplorer.util.NotificationUtils
import ca.pkay.rcloneexplorer.workmanager.SyncWorker
import ca.pkay.rcloneexplorer.workmanager.SyncWorker.Companion.EXTRA_TASK_ID
import java.util.UUID

class SyncServiceNotifications(var mContext: Context) {


    companion object {
        const val CHANNEL_ID = "ca.pkay.rcexplorer.sync_service"
        const val CHANNEL_SUCCESS_ID = "ca.pkay.rcexplorer.sync_service_success"
        const val CHANNEL_FAIL_ID = "ca.pkay.rcexplorer.sync_service_fail"


        const val PERSISTENT_NOTIFICATION_ID_FOR_SYNC = 162
        const val CANCEL_ID_NOTSET = "CANCEL_ID_NOTSET"
        const val TAG = "SyncServiceNotifications"

    }

    private var mReportManager = ReportNotifications(mContext)

    private val OPERATION_FAILED_GROUP = "ca.pkay.rcexplorer.OPERATION_FAILED_GROUP"
    private val OPERATION_SUCCESS_GROUP = "ca.pkay.rcexplorer.OPERATION_SUCCESS_GROUP"


    private var mCancelUnsetId: UUID = UUID.randomUUID()
    private var mCancelId: UUID = mCancelUnsetId

    fun setCancelId(id: UUID) {
        mCancelId = id
    }

    private fun useReports(): Boolean {
        val mSharedPreferences = PreferenceManager.getDefaultSharedPreferences(mContext)
        return mSharedPreferences.getBoolean(mContext.getString(R.string.pref_key_app_notification_reports), true)
    }

    fun showFailedNotificationOrReport(
        title: String,
        content: String,
        notificationId: Int,
        taskid: Long
    ) {
        if(!useReports()){
            showFailedNotification(content, notificationId, taskid)
            return
        }
        if(mReportManager.getFailures()<=1) {
            showFailedNotification(content, notificationId, taskid)
            mReportManager.lastFailedNotification(notificationId)
            mReportManager.addToFailureReport(title, content)
        } else {
            mReportManager.cancelLastFailedNotification()
            mReportManager.showFailReport(title, content)
        }
    }

    fun showFailedNotification(
        content: String,
        notificationId: Int,
        taskid: Long
    ) {
        val i = Intent(mContext, SyncRestartAction::class.java)
        i.putExtra(EXTRA_TASK_ID, taskid)

        val retryPendingIntent = PendingIntent.getService(mContext, taskid.toInt(), i, GenericSyncNotification.getFlags())
        val builder = NotificationCompat.Builder(mContext, CHANNEL_FAIL_ID)
            .setSmallIcon(R.drawable.ic_twotone_cloud_error_24)
            .setContentTitle(mContext.getString(R.string.operation_failed))
            .setContentText(content)
            .setStyle(
                NotificationCompat.BigTextStyle().bigText(content)
            )
            .setGroup(OPERATION_FAILED_GROUP)
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .addAction(
                R.drawable.ic_refresh,
                mContext.getString(R.string.retry_failed_sync),
                retryPendingIntent
            )
        NotificationUtils.createNotification(mContext, notificationId, builder.build())
    }
    fun showCancelledNotificationOrReport(
        content: String,
        notificationId: Int,
        taskid: Long) {

        if(!useReports()){
            showCancelledNotification(content, notificationId, taskid)
            return
        }
        var title = mContext.getString(R.string.operation_failed_cancelled)
        if(mReportManager.getFailures()<=1) {
            showCancelledNotification(content, notificationId, taskid)
            mReportManager.lastFailedNotification(notificationId)
            mReportManager.addToFailureReport(title, content)
        } else {
            mReportManager.cancelLastFailedNotification()
            mReportManager.showFailReport(title, content)
        }
    }

    fun showCancelledNotification(
        content: String,
        notificationId: Int,
        taskid: Long
    ) {
        val i = Intent(mContext, SyncRestartAction::class.java)
        i.putExtra(EXTRA_TASK_ID, taskid)

        val retryPendingIntent = PendingIntent.getService(mContext, taskid.toInt(), i, GenericSyncNotification.getFlags())
        val builder = NotificationCompat.Builder(mContext, CHANNEL_FAIL_ID)
            .setSmallIcon(R.drawable.ic_twotone_cloud_error_24)
            .setContentTitle(mContext.getString(R.string.operation_failed_cancelled))
            .setContentText(content)
            .setStyle(
                NotificationCompat.BigTextStyle().bigText(content)
            )
            .setGroup(OPERATION_FAILED_GROUP)
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .addAction(
                R.drawable.ic_refresh,
                mContext.getString(R.string.retry_failed_sync),
                retryPendingIntent
            )
        NotificationUtils.createNotification(mContext, notificationId, builder.build())
    }

    fun showSuccessNotificationOrReport(
        title: String,
        content: String,
        notificationId: Int
    ) {

        if(!useReports()){
            showSuccessNotification(title, content, notificationId)
            return
        }

        if(mReportManager.getSucesses()<=1) {
            showSuccessNotification(title, content, notificationId)
            mReportManager.lastSuccededNotification(notificationId)
            mReportManager.addToSuccessReport(title, content)
        } else {
            mReportManager.cancelLastSuccededNotification()
            mReportManager.showSuccessReport(title, content)
        }
    }
    fun showSuccessNotification(title: String, content: String, notificationId: Int) {
        val builder = NotificationCompat.Builder(mContext, CHANNEL_SUCCESS_ID)
            .setSmallIcon(R.drawable.ic_twotone_cloud_done_24)
            .setContentTitle(mContext.getString(R.string.operation_success, title))
            .setContentText(content)
            .setStyle(
                NotificationCompat.BigTextStyle().bigText(
                    content
                )
            )
            .setGroup(OPERATION_SUCCESS_GROUP)
            .setPriority(NotificationCompat.PRIORITY_LOW)
        NotificationUtils.createNotification(mContext, notificationId, builder.build())
    }

    @Deprecated("Use with specific notification id")
    fun updateSyncNotification(
        title: String,
        content: String,
        bigTextArray: ArrayList<String>,
        percent: Int
    ): Notification? {
        return updateSyncNotification(
            title,
            content,
            bigTextArray,
            percent,
            PERSISTENT_NOTIFICATION_ID_FOR_SYNC
        )
    }

    fun updateSyncNotification(
        title: String,
        content: String,
        bigTextArray: ArrayList<String>,
        percent: Int,
        notificationId: Int
    ): Notification? {
        if(content.isBlank()){
            FLog.e(TAG, "Missing notification content!")
            return null
        }

        val builder = GenericSyncNotification(mContext).updateGenericNotification(
            mContext.getString(R.string.syncing_service, title),
            content,
            R.drawable.ic_twotone_rounded_cloud_sync_24,
            bigTextArray,
            percent,
            SyncWorker::class.java,
            null,
            CHANNEL_ID
        )

        if(mCancelId != mCancelUnsetId) {

            val intent = WorkManager.getInstance(mContext)
                .createCancelPendingIntent(mCancelId)

            builder.clearActions()
            builder.addAction(
                R.drawable.ic_cancel_download,
                mContext.getString(R.string.cancel),
                intent
            )
        }

        val notification = builder.build()
        NotificationUtils.createNotification(mContext, notificationId, notification)
        return notification
    }

    fun cancelSyncNotification(notificationId: Int) {
        val notificationManagerCompat = NotificationManagerCompat.from(mContext)
        notificationManagerCompat.cancel(notificationId)
    }
}
